# Client getINNOtized 
#### Sales data was collected for each month in the entire year of 2019. 


### 1. Business Understanding
We are assigned a client by getINNOtized to design and deliver an end-to-end business  intelligence solution. Our client has collected transactional data for the year 2019 but hasn’t been able to  put it to good use. 
Sales data was collected for each month in the entire year of 2019. 
The data for the first half of the year (January to June) was collected in excel and saved as csv files before management decided to use databases to store their data.

### Objective
 The client hopes you can analyze the data and put together a report  to help them find opportunities to drive more sales and work more efficiently. 

 ### **Questions**
1. How much money did we make this year? 

2. Can we identify any seasonality in the  sales? 

3. What are our best and worst-selling products? 

4. How do sales compare to previous months or weeks? 

5. Which cities are our products delivered to most? 

6. How do product categories compare in revenue generated and quantities  ordered? 

7. You are required to show additional details from your findings in your data.

#### 2. Data Understanding
Here, we explored and familiarized with data the client by getINNOtized from 2019 . This involved examining the structure, quality, and content of the dataset. We further identified the relevant variables, data types, and potential issues, such as missing values or outliers.

## Installations

Packages for accessing remote database with data

In [ ]:

import pyodbc 
from dotenv import dotenv_values #import the dotenv_values function from the dotenv package
import warnings 
warnings.filterwarnings('ignore')

# Data handling

import pandas as pd
import numpy as np

# Data Visualisation

import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
import plotly.colors as colors

In [4]:
import zipfile
import os

# path to the file zip
zip_file_path = "OneDrive_1_2-9-2024.zip"

# Folder 
extract_to_dir = "data"

# Create folder if not exist
if not os.path.exists(extract_to_dir):
    os.makedirs(extract_to_dir)

# Unzip file
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_to_dir)

print("file unzip sucessfully.")


Fichier zip décompressé avec succès.


### Load Data

In [2]:
#Read data january
data_jan =pd.read_csv('data/Sales_January_2019.csv')
data_jan.head()

Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [3]:
data_jan.describe().T

count unique                   top  freq
Order ID          9697   9269              Order ID    16
Product           9697     20  USB-C Charging Cable  1171
Quantity Ordered  9697      8                     1  8795
Price Each        9697     19                 11.95  1171
Order Date        9697   8077            Order Date    16
Purchase Address  9697   9161      Purchase Address    16

In [4]:
data_jan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9723 entries, 0 to 9722
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          9697 non-null   object
 1   Product           9697 non-null   object
 2   Quantity Ordered  9697 non-null   object
 3   Price Each        9697 non-null   object
 4   Order Date        9697 non-null   object
 5   Purchase Address  9697 non-null   object
dtypes: object(6)
memory usage: 455.9+ KB


In [5]:
data_jan.isna().sum()

Order ID            26
Product             26
Quantity Ordered    26
Price Each          26
Order Date          26
Purchase Address    26
dtype: int64

In [6]:
#let's check the dimension of our data januray
data_jan.shape

(9723, 6)

In [7]:
data_jan.columns

Index(['Order ID', 'Product', 'Quantity Ordered', 'Price Each', 'Order Date',
       'Purchase Address'],
      dtype='object')

In [8]:
#Read data February
data_feb =pd.read_csv('data/Sales_February_2019.csv')
data_feb.head()

Order ID                   Product Quantity Ordered Price Each  \
0   150502                    iPhone                1        700   
1   150503     AA Batteries (4-pack)                1       3.84   
2   150504    27in 4K Gaming Monitor                1     389.99   
3   150505  Lightning Charging Cable                1      14.95   
4   150506     AA Batteries (4-pack)                2       3.84   

       Order Date                     Purchase Address  
0  02/18/19 01:35    866 Spruce St, Portland, ME 04101  
1  02/13/19 07:24  18 13th St, San Francisco, CA 94016  
2  02/18/19 09:46   52 6th St, New York City, NY 10001  
3  02/02/19 16:47     129 Cherry St, Atlanta, GA 30301  
4  02/28/19 20:32    548 Lincoln St, Seattle, WA 98101

In [9]:
data_feb.shape

(12036, 6)

In [10]:
#Read data February
data_March =pd.read_csv('data/Sales_March_2019.csv')
data_March.head()

Order ID                     Product Quantity Ordered Price Each  \
0   162009                      iPhone                1        700   
1   162009    Lightning Charging Cable                1      14.95   
2   162009            Wired Headphones                2      11.99   
3   162010  Bose SoundSport Headphones                1      99.99   
4   162011      34in Ultrawide Monitor                1     379.99   

       Order Date                      Purchase Address  
0  03/28/19 20:59       942 Church St, Austin, TX 73301  
1  03/28/19 20:59       942 Church St, Austin, TX 73301  
2  03/28/19 20:59       942 Church St, Austin, TX 73301  
3  03/17/19 05:39  261 10th St, San Francisco, CA 94016  
4  03/10/19 00:01  764 13th St, San Francisco, CA 94016

In [11]:
data_March.shape

(15226, 6)

In [12]:
#Read data April
data_April =pd.read_csv('data/Sales_April_2019.csv')
data_April.head()

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   
4   176560            Wired Headphones                1      11.99   

       Order Date                      Purchase Address  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001  
1             NaN                                   NaN  
2  04/07/19 22:30     682 Chestnut St, Boston, MA 02215  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001  
4  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001

In [13]:
data_April.shape

(18383, 6)

In [14]:
#load data May
data_May = pd.read_csv('data/Sales_May_2019.csv')
data_May.head()

Order ID                 Product Quantity Ordered Price Each  \
0   194095        Wired Headphones                1      11.99   
1   194096   AA Batteries (4-pack)                1       3.84   
2   194097        27in FHD Monitor                1     149.99   
3   194098        Wired Headphones                1      11.99   
4   194099  AAA Batteries (4-pack)                2       2.99   

       Order Date                         Purchase Address  
0  05/16/19 17:14      669 2nd St, New York City, NY 10001  
1  05/19/19 14:43          844 Walnut St, Dallas, TX 75001  
2  05/24/19 11:36  164 Madison St, New York City, NY 10001  
3  05/02/19 20:40          622 Meadow St, Dallas, TX 75001  
4  05/11/19 22:55          17 Church St, Seattle, WA 98101

In [15]:
data_May.shape

(16635, 6)

In [28]:
#load June data
data_jun = pd.read_csv('data/Sales_June_2019.csv')
data_jan.head()

Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [29]:
data_July = pd.read_csv('data/Sales_July_2019.csv')
data_July.head()

Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

                    Order_Date                  Purchase_Address  
0  2026-07-19 16:51:00.0000000   389 South St, Atlanta, GA 30301  
1  2005-07-19 08:55:00.0000000     590 4th St, Seattle, WA 98101  
2  2029-07-19 12:41:00.0000000    861 Hill St, Atlanta, GA 30301  
3  2028-07-19 10:15:00.0000000   190 Ridge St, Atlanta, GA 30301  
4  2031-07-19 02:13:00.0000000  824 Forest St, Seattle, WA 98101

In [18]:
data_Aug = pd.read_csv('data/Sales_August_2019.csv')
data_Aug.shape

(12011, 6)

In [19]:
data_Aug.head()

Order_ID                     Product  Quantity_Ordered  Price_Each  \
0  236670.0            Wired Headphones               2.0   11.990000   
1  236671.0  Bose SoundSport Headphones               1.0   99.989998   
2  236672.0                      iPhone               1.0  700.000000   
3  236673.0       AA Batteries (4-pack)               2.0    3.840000   
4  236674.0       AA Batteries (4-pack)               2.0    3.840000   

                    Order_Date                      Purchase_Address  
0  2031-08-19 22:21:00.0000000      359 Spruce St, Seattle, WA 98101  
1  2015-08-19 15:11:00.0000000        492 Ridge St, Dallas, TX 75001  
2  2006-08-19 14:40:00.0000000        149 7th St, Portland, OR 97035  
3  2029-08-19 20:59:00.0000000     631 2nd St, Los Angeles, CA 90001  
4  2015-08-19 19:53:00.0000000  736 14th St, New York City, NY 10001

In [20]:
data_Sept = pd.read_csv('data/Sales_September_2019.csv')
data_Sept.shape

(11686, 6)

In [21]:
data_Oct = pd.read_csv('data/Sales_October_2019.csv')
data_Oct.head()

Order_ID                 Product  Quantity_Ordered  Price_Each  \
0  259358.0  34in Ultrawide Monitor               1.0  379.989990   
1  259359.0  27in 4K Gaming Monitor               1.0  389.989990   
2  259360.0  AAA Batteries (4-pack)               2.0    2.990000   
3  259361.0        27in FHD Monitor               1.0  149.990005   
4  259362.0        Wired Headphones               1.0   11.990000   

                    Order_Date                           Purchase_Address  
0  2028-10-19 10:56:00.0000000            609 Cherry St, Dallas, TX 75001  
1  2028-10-19 17:26:00.0000000          225 5th St, Los Angeles, CA 90001  
2  2024-10-19 17:20:00.0000000       967 12th St, New York City, NY 10001  
3  2014-10-19 22:26:00.0000000  628 Jefferson St, New York City, NY 10001  
4  2007-10-19 16:10:00.0000000         534 14th St, Los Angeles, CA 90001

In [22]:
data_Oct.shape

(20379, 6)

In [23]:
#load data November
data_Nov = pd.read_csv('data/Sales_November_2019.csv')
data_Nov.head()

Order_ID                     Product  Quantity_Ordered  Price_Each  \
0  278797.0            Wired Headphones               1.0   11.990000   
1  278798.0        USB-C Charging Cable               2.0   11.950000   
2  278799.0    Apple Airpods Headphones               1.0  150.000000   
3  278800.0            27in FHD Monitor               1.0  149.990005   
4  278801.0  Bose SoundSport Headphones               1.0   99.989998   

                    Order_Date                      Purchase_Address  
0  2021-11-19 09:54:00.0000000   46 Park St, New York City, NY 10001  
1  2017-11-19 10:03:00.0000000      962 Hickory St, Austin, TX 73301  
2  2019-11-19 14:56:00.0000000  464 Cherry St, Los Angeles, CA 90001  
3  2025-11-19 22:24:00.0000000        649 10th St, Seattle, WA 98101  
4  2009-11-19 13:56:00.0000000         522 Hill St, Boston, MA 02215

In [24]:
data_dec = pd.read_csv('data/Sales_December_2019.csv')
data_dec.head()

Order_ID               Product  Quantity_Ordered   Price_Each  \
0  295665.0    Macbook Pro Laptop               1.0  1700.000000   
1  295666.0    LG Washing Machine               1.0   600.000000   
2  295667.0  USB-C Charging Cable               1.0    11.950000   
3  295668.0      27in FHD Monitor               1.0   149.990005   
4  295669.0  USB-C Charging Cable               1.0    11.950000   

                    Order_Date                        Purchase_Address  
0  2030-12-19 00:01:00.0000000  136 Church St, New York City, NY 10001  
1  2029-12-19 07:03:00.0000000     562 2nd St, New York City, NY 10001  
2  2012-12-19 18:21:00.0000000    277 Main St, New York City, NY 10001  
3  2022-12-19 15:13:00.0000000     410 6th St, San Francisco, CA 94016  
4  2018-12-19 12:38:00.0000000           43 Hill St, Atlanta, GA 30301

In [30]:
data_dec.head()

Order_ID               Product  Quantity_Ordered   Price_Each  \
0  295665.0    Macbook Pro Laptop               1.0  1700.000000   
1  295666.0    LG Washing Machine               1.0   600.000000   
2  295667.0  USB-C Charging Cable               1.0    11.950000   
3  295668.0      27in FHD Monitor               1.0   149.990005   
4  295669.0  USB-C Charging Cable               1.0    11.950000   

                    Order_Date                        Purchase_Address  
0  2030-12-19 00:01:00.0000000  136 Church St, New York City, NY 10001  
1  2029-12-19 07:03:00.0000000     562 2nd St, New York City, NY 10001  
2  2012-12-19 18:21:00.0000000    277 Main St, New York City, NY 10001  
3  2022-12-19 15:13:00.0000000     410 6th St, San Francisco, CA 94016  
4  2018-12-19 12:38:00.0000000           43 Hill St, Atlanta, GA 30301

##### Data Merge


In [32]:
data_six_last_mont =pd.concat ([data_July,data_Aug,data_Sept,data_Nov,data_dec],ignore_index=False)
data_six_last_mont.head()

Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

                    Order_Date                  Purchase_Address  
0  2026-07-19 16:51:00.0000000   389 South St, Atlanta, GA 30301  
1  2005-07-19 08:55:00.0000000     590 4th St, Seattle, WA 98101  
2  2029-07-19 12:41:00.0000000    861 Hill St, Atlanta, GA 30301  
3  2028-07-19 10:15:00.0000000   190 Ridge St, Atlanta, GA 30301  
4  2031-07-19 02:13:00.0000000  824 Forest St, Seattle, WA 98101

In [33]:
data_six_first_month = pd.concat([data_jan,data_feb,data_March,data_April,data_May,data_jun],ignore_index=True)
data_six_first_month.head()

Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [34]:
#check columns six first month
data_six_first_month.columns

Index(['Order ID', 'Product', 'Quantity Ordered', 'Price Each', 'Order Date',
       'Purchase Address'],
      dtype='object')

In [35]:
#let's check data sixe last month
data_six_last_mont.columns

Index(['Order_ID', 'Product', 'Quantity_Ordered', 'Price_Each', 'Order_Date',
       'Purchase_Address'],
      dtype='object')

#### Rename Column data six last months
We noticed that for the columns of the first 6 months are different from that of the last 6 months. Let's try to rename that of the last 6 months in order to confirm the data.

In [45]:
df1 = pd.DataFrame(data_six_last_mont)
df1 = df1.rename(columns={
    'Order_ID':'Order ID',
    'Quantity_Ordered':'Quantity Ordered',
    'Price_Each':'Price Each',
    'Order_Date':'Order Date',
    'Purchase_Address':'Purchase Address'
})

df1.head()

Order ID                   Product  Quantity Ordered  Price Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

                    Order Date                  Purchase Address  
0  2026-07-19 16:51:00.0000000   389 South St, Atlanta, GA 30301  
1  2005-07-19 08:55:00.0000000     590 4th St, Seattle, WA 98101  
2  2029-07-19 12:41:00.0000000    861 Hill St, Atlanta, GA 30301  
3  2028-07-19 10:15:00.0000000   190 Ridge St, Atlanta, GA 30301  
4  2031-07-19 02:13:00.0000000  824 Forest St, Seattle, WA 98101

In [46]:
#let's drop our Order ID
df1.drop(['Order ID'],axis=1,inplace=True)
df1.head()

Product  Quantity Ordered  Price Each  \
0  Apple Airpods Headphones               1.0      150.00   
1             Flatscreen TV               1.0      300.00   
2     AA Batteries (4-pack)               1.0        3.84   
3     AA Batteries (4-pack)               1.0        3.84   
4    AAA Batteries (4-pack)               5.0        2.99   

                    Order Date                  Purchase Address  
0  2026-07-19 16:51:00.0000000   389 South St, Atlanta, GA 30301  
1  2005-07-19 08:55:00.0000000     590 4th St, Seattle, WA 98101  
2  2029-07-19 12:41:00.0000000    861 Hill St, Atlanta, GA 30301  
3  2028-07-19 10:15:00.0000000   190 Ridge St, Atlanta, GA 30301  
4  2031-07-19 02:13:00.0000000  824 Forest St, Seattle, WA 98101

In [52]:
#let's drop our columns for data six irst month
data_six_first_month.drop(['Order ID'],axis=1,inplace=True)
data_six_first_month.head()

Product Quantity Ordered Price Each      Order Date  \
0                    iPhone                1        700  01/22/19 21:25   
1  Lightning Charging Cable                1      14.95  01/28/19 14:15   
2          Wired Headphones                2      11.99  01/17/19 13:33   
3          27in FHD Monitor                1     149.99  01/05/19 20:33   
4          Wired Headphones                1      11.99  01/25/19 11:59   

                        Purchase Address  
0        944 Walnut St, Boston, MA 02215  
1       185 Maple St, Portland, OR 97035  
2  538 Adams St, San Francisco, CA 94016  
3     738 10th St, Los Angeles, CA 90001  
4          387 10th St, Austin, TX 73301

#### Merge All Data 
Now we will merge the data in order to have a unique dataset

In [53]:
data = pd.concat([df1,data_six_first_month],ignore_index=False)
data.head()

Product Quantity Ordered Price Each  \
0  Apple Airpods Headphones              1.0      150.0   
1             Flatscreen TV              1.0      300.0   
2     AA Batteries (4-pack)              1.0       3.84   
3     AA Batteries (4-pack)              1.0       3.84   
4    AAA Batteries (4-pack)              5.0       2.99   

                    Order Date                  Purchase Address  
0  2026-07-19 16:51:00.0000000   389 South St, Atlanta, GA 30301  
1  2005-07-19 08:55:00.0000000     590 4th St, Seattle, WA 98101  
2  2029-07-19 12:41:00.0000000    861 Hill St, Atlanta, GA 30301  
3  2028-07-19 10:15:00.0000000   190 Ridge St, Atlanta, GA 30301  
4  2031-07-19 02:13:00.0000000  824 Forest St, Seattle, WA 98101

#### 2.Data Cleaning

In [54]:
data.tail()

Product Quantity Ordered Price Each      Order Date  \
85620  AAA Batteries (4-pack)                1       2.99  06/07/19 19:02   
85621        27in FHD Monitor                1     149.99  06/01/19 19:29   
85622    USB-C Charging Cable                1      11.95  06/22/19 18:57   
85623    USB-C Charging Cable                1      11.95  06/26/19 18:35   
85624  AAA Batteries (4-pack)                1       2.99  06/25/19 14:33   

                            Purchase Address  
85620          795 Pine St, Boston, MA 02215  
85621  495 North St, New York City, NY 10001  
85622  319 Ridge St, San Francisco, CA 94016  
85623   916 Main St, San Francisco, CA 94016  
85624         209 11th St, Atlanta, GA 30301

In [55]:
#let's check describe our new dataset
data.describe()

Product Quantity Ordered Price Each  Order Date  \
count                 165988           165826     165826      165826   
unique                    20               18         40      127549   
top     USB-C Charging Cable                1      11.95  Order Date   
freq                   19460            77205       9952         160   

        Purchase Address  
count             165988  
unique            128509  
top     Purchase Address  
freq                 322

In [56]:
#let's check shape
data.shape

(166471, 5)

Our new dataset now has 5 columns and 166471 rows. On the other hand, in our describe we don't see the samething, this is certainly due to the presence of missing values.

###### 2.1 MISSING VALUE

In [57]:
#let's check if we are missing and what column
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 166471 entries, 0 to 85624
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Product           165988 non-null  object
 1   Quantity Ordered  165826 non-null  object
 2   Price Each        165826 non-null  object
 3   Order Date        165826 non-null  object
 4   Purchase Address  165988 non-null  object
dtypes: object(5)
memory usage: 7.6+ MB


In [58]:
data.isna().sum()

Product             483
Quantity Ordered    645
Price Each          645
Order Date          645
Purchase Address    483
dtype: int64

Apparently all our 5 columns contain missing values, let's try to see more of what these missing values hide by column.

In [67]:
#let's check our Product column
data['Product'].value_counts()

Product
USB-C Charging Cable          19460
Lightning Charging Cable      19238
AAA Batteries (4-pack)        18402
AA Batteries (4-pack)         18337
Wired Headphones              16787
Apple Airpods Headphones      13884
Bose SoundSport Headphones    11885
27in FHD Monitor               6716
iPhone                         6127
27in 4K Gaming Monitor         5511
34in Ultrawide Monitor         5510
Google Phone                   4928
Flatscreen TV                  4276
Macbook Pro Laptop             4199
20in Monitor                   3674
ThinkPad Laptop                3669
Vareebadd Phone                1862
LG Washing Machine              605
LG Dryer                        596
Product                         322
Name: count, dtype: int64

: 